addons/website_sale/controllers/main.py

```diff
@@ -1131,6 +1131,12 @@ def shop_payment_validate(self, transaction_id=None, sale_order_id=None, **post)
             order = request.env['sale.order'].sudo().browse(sale_order_id)
             assert order.id == request.session.get('sale_last_order_id')
 
+        render_values = self._get_shop_payment_values(order, **post)
```

Comment:
It is better to repeat only the check instead of redoing all the logic to create render values that are never used.
```
has_storable_products = any(line.product_id.type in ['consu', 'product'] for line in order.order_line)
if not order._get_delivery_methods() and has_storable_products:
     raise ValidationError(_('Unable to validate your order.'))
```

Reply:
Hello! The thing is that the error can come from another module and I wanted to block it in a more generic way.

Ex:
https://github.com/odoo-dev/enterprise/blob/6bfa82866e86963050decea9841d2ef1c7a72185/website_sale_account_avatax/controllers/main.py#L9-L22

I could extract those errors in a submethod that i would call `_get_shop_payment_errors` in all those modules so I can use them without the render logic but the diff will be a bit bigger.

Reply:
Oh, you are right!
It would be great for the future to indeed do a bit of refactoring and extract the logic as you suggest.

Reply:
Yes, i think now it makes sense to separate your commit. I will help with resolving conflicts in 17 template. Could you please change the fw up to master @AntoineVDV ?